### PR TITLE
Allow the creation of private events 

### LINKED ISSUE #982: Add private boolean to events
Either a boolean or a code that allows you to publish an event but it doesn't show up on the public index endpoints. You must have the event_id or the code (if we go that route) to view the event.

### PR DESCRIPTION
**Pull request by SWvheerden**
### Closes Issues:
fixes #982 
### Description:
This hides events that are private and allows the user to add give an access code to view the event
### Migrations
add private code to events table

### Environment Variables
 * No change

### CODE REVIEW

File: Cargo.lock [hunk 1]
```diff
@@ -1,3 +1,5 @@
+# This file is automatically @generated by Cargo.
```
Comment (Krakaw) [inline Cargo.lock]: Don't commit this file.
Use
```bash
git update-index --assume-unchanged ./Cargo.lock 
```

File: db/src/models/events.rs [hunk 2]
```diff
@@ -397,12 +399,26 @@ impl Event {
...
-        )
+    pub fn find(
+        id: Uuid,
+        user: Option<User>,
+        private_code: Option<String>,
+        conn: &PgConnection,
+    ) -> Result<Event, DatabaseError> {
+        let mut query = events::table.find(id).into_boxed();
+        match user {
+            Some(user) => {if !user.get_global_scopes().contains(&Scopes::OrgAdmin) {
```
Comment (Krakaw) [inline db/src/models/events.rs]: This can be global scopes, but also organization scopes. Basically anyone that is part of the org.

File: api/src/controllers/events.rs [hunk 3]
```diff
@@ -544,7 +561,7 @@ pub fn show_from_organizations(
 ) -> Result<WebPayload<EventSummaryResult>, BigNeonError> {
     let conn = connection.get();
     let org = Organization::find(path.id, conn)?;
-    user.requires_scope_for_organization(Scopes::RedeemTicket, &org, conn)?;
+    user.requires_scope_for_organization(Scopes::CanViewPrivate, &org, conn)?;
```
Comment (mikethetike) [inline api/src/controllers/events.rs]: I don't think this is the correct scope. Perhaps `EventWrite || RedeemTicket` (For scanning or editing)

File: api/tests/functional/base/events.rs [hunk 4]
```diff
@@ -104,7 +104,7 @@ pub fn show_box_office_pricing(role: Roles, should_test_succeed: bool) {
 
     if should_test_succeed {
         let event_expected_json =
-            expected_show_json(event, organization, venue, true, None, None, conn);
+            expected_show_json(event, organization, venue, true, None, None, conn, 1);
```
Comment (mikethetike) [inline api/tests/functional/base/events.rs]: Is this change related to this PR? Seems to have snuck in

File: api/tests/functional/base/events.rs [hunk 5]
```diff
@@ -690,6 +690,7 @@ pub fn expected_show_json(
     redemption_code: Option<String>,
     filter_ticket_type_ids: Option<Vec<Uuid>>,
     connection: &PgConnection,
+    intrested_users: u32,
```
Comment (mikethetike) [inline api/tests/functional/base/events.rs]: spelling

File: db/src/models/events.rs [hunk 6]
```diff
@@ -938,6 +941,8 @@ impl Event {
         match user {
             Some(user) => {
                 // Admin results include all drafts across organizations
+                println!("we have user: {:?}", user);
```
Comment (mikethetike) [inline db/src/models/events.rs]: Debug code?

File: db/src/models/scopes.rs [hunk 7]
```diff
@@ -14,6 +14,7 @@ pub enum Scopes {
     CodeWrite,
     CompRead,
     CompWrite,
+    CanViewPrivate,
```
Comment (mikethetike) [inline db/src/models/scopes.rs]: We can probably infer this from the other scopes. If a user can edit events in an org, they can see private events. If they need to scan, they can see private events

File: api/tests/functional/base/events.rs (hunk 4 above)
Comment (SWvheerden) [reply api/tests/functional/base/events.rs]: Yes it is, I needed to change the function to allow the passing in the number of interested users.  This is so I can reuse the function.

File: api/src/controllers/events.rs (hunk 3 above)
Comment (SWvheerden) [reply api/src/controllers/events.rs]: Purposely changed the scope. Boxoffice does not have the right redeemticket.

File: db/src/models/scopes.rs (hunk 7 above)
Comment (SWvheerden) [reply db/src/models/scopes.rs]: This is the only shared right between boxoffice, doorman, org member.

File: db/src/models/events.rs [hunk 8]
```diff
@@ -950,7 +957,7 @@ impl Event {
                                 .le(dsl::now.nullable())
                                 .or(events::status.ne(EventStatus::Published))
                                 .or(organization_users::user_id.eq(user.id)),
-                        );
```
Comment (sdbondi) [inline db/src/models/events.rs]: `;` should be here

**Comment by mikethetike** [comment]
@SWvheerden Just need to address the comments by @Krakaw and then this can go in

File: db/src/models/events.rs [hunk 9]
```diff
@@ -170,6 +171,7 @@ pub struct EventEditableAttributes {
     #[serde(default, deserialize_with = "double_option_deserialize_unless_blank")]
     pub override_status: Option<Option<EventOverrideStatus>>,
     pub event_end: Option<NaiveDateTime>,
+    pub private_access_code: Option<String>,
```
Comment (mikethetike) [inline db/src/models/events.rs]: This probably needs to be a double option to allow setting it to null. See `external_url`

**Comment by SWvheerden** [comment]
@Krakaw 

On your comments,
All is done except the limits, should there be a limit? and if, what should it be

**Comment by Krakaw** [comment]
> @Krakaw
> 
> On your comments,
> All is done except the limits, should there be a limit? and if, what should it be

Make it 6 chars [a-z0-9]{6} if you're lowercasing it to compare.
